File: skills/java-mysql-query/scripts/csv-exporter.js
#!/usr/bin/env node
/**
 * CSV Exporter (Node.js 版)
 * 将 SQL 查询结果导出为 CSV 文件。
 * 用法: node csv-exporter.js --db mydb "SELECT * FROM user" --output users.csv
 */
const fs = require("fs");
const { execSync } = require("child_process");
const { MySQLQuery } = require("./database-query.js");
const path = require("path");

function toCsv(data) {
  if (!data || data.length === 0) return "";
  const headers = Object.keys(data[0]);
  const lines = [headers.map(h => `"${h.replace(/"/g, '""')}"`).join(",")];
  data.forEach(row => {
    lines.push(headers.map(h => `"${String(row[h] || "").replace(/"/g, '""')}"`).join(","));
  });
  return lines.join("\n");
}

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { host: "localhost", port: "3306", user: "root", output: "export.csv", ssl: "false" };
  let sql = "";
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--host": opts.host = args[++i]; break;
      case "--port": opts.port = args[++i]; break;
      case "--db": opts.db = args[++i]; break;
      case "--user": opts.user = args[++i]; break;
      case "--password": opts.password = args[++i]; break;
      case "--ssl": opts.ssl = args[++i]; break;
      case "--output": case "-o": opts.output = args[++i]; break;
      case "--input": opts.input = args[++i]; break;
      default: sql = (sql ? sql + " " : "") + args[i];
    }
  }
  opts.sql = sql;
  return opts;
}

function run(opts) {
  if (opts.input) {
    const raw = fs.readFileSync(opts.input, "utf-8");
    const data = JSON.parse(raw);
    const arr = Array.isArray(data) ? data : (data.data || [data]);
    const csv = toCsv(arr);
    fs.writeFileSync(opts.output, "\ufeff" + csv, "utf-8");
    console.log(JSON.stringify({ status: "success", output: opts.output, rows: arr.length }));
  } else if (opts.sql) {
    const cmd = `python database_query.py --host ${opts.host} --port ${opts.port} --db ${opts.db} --user ${opts.user}${opts.password ? " --password " + opts.password : ""} --ssl ${opts.ssl} --export-csv "${opts.sql.replace(/"/g, '\\"')}" --output "${opts.output}"`;
    try {
      const out = execSync(cmd, { timeout: 60000, shell: true }).toString();
      console.log(out.trim());
    } catch (e) { console.error(JSON.stringify({ status: "error", message: e.message })); }
  } else {
    console.log("用法: node csv-exporter.js --db mydb [--output file.csv] \"SELECT * FROM table\"");
  }
}
run(parseArgs());

File: skills/java-mysql-query/scripts/csv_exporter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
CSV Exporter (Python 版)
将 SQL 查询结果导出为 CSV 文件。
用法：
  python csv_exporter.py --host localhost --db mydb "SELECT * FROM user" --output users.csv
  python csv_exporter.py --input result.json --output output.csv
"""
import json, csv, sys, os, argparse
from database_query import MySQLQuery

def export_csv(data, output_path):
    if not data: return 0
    headers = list(data[0].keys()) if isinstance(data[0], dict) else []
    with open(output_path, "w", newline="", encoding="utf-8-sig") as f:
        writer = csv.writer(f)
        writer.writerow(headers)
        for row in data:
            writer.writerow([str(row.get(h, "")) for h in headers] if isinstance(row, dict) else row)
    return len(data)

def main():
    parser = argparse.ArgumentParser(description="CSV Exporter 导出工具")
    parser.add_argument("--host", default="localhost"); parser.add_argument("--port", type=int, default=3306)
    parser.add_argument("--db", required=True); parser.add_argument("--user", default="root")
    parser.add_argument("--password", default=os.environ.get("DB_PASSWORD", ""))
    parser.add_argument("--ssl", default="false")
    parser.add_argument("--input", help="已有JSON结果文件")
    parser.add_argument("sql", nargs="?", help="SQL查询语句")
    parser.add_argument("--output", "-o", default="export.csv", help="输出CSV文件路径")
    args = parser.parse_args()
    if args.input:
        with open(args.input) as f: data = json.load(f)
        if isinstance(data, dict): data = data.get("data", data)
        count = export_csv(data, args.output)
        print(json.dumps({"status":"success","output":args.output,"rows":count}))
    elif args.sql:
        q = MySQLQuery(args.host, args.port, args.db, args.user, args.password, args.ssl)
        q.connect()
        result = q.export_csv(args.sql, args.output)
        q.close()
        print(json.dumps(result, ensure_ascii=False))
    else:
        parser.print_help()
if __name__ == "__main__": main()
